### ISSUE TITLE
Lock

### ISSUE THREAD

**Issue by KTibow** [bachya/simplisafe-python#47]
Recently, SimpliSafe released a new lock. You can see it at [https://simplisafe.com/smart-lock](https://simplisafe.com/smart-lock). Anyway, soon it would be nice to support this feature. There are some nice, even free apps you can install and they will analyze network traffic, even over https. In order to stay up to date, a lock should be added as a device you can lock+unlock.
(Of course, you would need somebody with that lock in order to analyze traffic.)
**Additional context**
Lock: [simplisafe lock](https://simplisafe.com/smart-lock)
Free Android packet capture app: [Packet Capture](https://play.google.com/store/apps/details?id=app.greyshirts.sslcapture)
SimpliSafe app: [simplisafe app](https://play.google.com/store/apps/details?id=com.simplisafe.mobile)

**Comment by bachya** [comment]
>(Of course, you would need somebody with that lock in order to analyze traffic.)

This is the key. I don't plan on buying one of these locks, so I will need someone who has one to assist in the traffic sniffing.

Will keep this open in the hopes that a partner appears!

**Comment by KTibow** [comment]
If all goes well, I should be getting one soon. It's just that some of my friends want to make sure that Packet Capture isn't stealing my data. I'll let you know as soon as I grab the packets.

**Comment by aceresia** [comment]
I have the lock and here is the JSON for the sensors call:

`{
		"serial": "x",
		"type": 16,
		"status": {
			"pinPadState": 0,
			"lockState": 1,
			"pinPadOffline": false,
			"pinPadLowBattery": false,
			"lockDisabled": false,
			"lockLowBattery": false,
			"calibrationErrDelta": 0,
			"calibrationErrZero": 0,
			"lockJamState": 0
		},
		"name": "Front Door",
		"deviceGroupID": 1,
		"firmwareVersion": "x",
		"bootVersion": "x",
		"setting": {
			"autoLock": 3,
			"away": 1,
			"home": 1,
			"awayToOff": 0,
			"homeToOff": 1
		},
		"flags": {
			"swingerShutdown": false,
			"lowBattery": false,
			"offline": false
		}
	}, {
		"serial": "x",
		"type": 253,
		"status": {},
		"name": "Front Door",
		"deviceGroupID": 1,
		"setting": {},
		"flags": {
			"swingerShutdown": false,
			"lowBattery": false,
			"offline": false
		}
	}`

I can get the lock and unlock commands when I get home.

**Comment by bachya** [comment]
@aceresia Thanks for checking in. That looks like a partial response payload; it would be helpful to see (a) the request URL and (b) the entire response.

**Comment by aceresia** [comment]
Sure, The URL is https://api.simplisafe.com/v1/ss3/subscriptions/<alarm_id>/sensors?forceUpdate=false

```{
	"account": "<account_number>",
	"success": true,
	"sensors": [{
		"serial": "<device_serial>",
		"type": 16,
		"status": {
			"pinPadState": 0,
			"lockState": 1,
			"pinPadOffline": false,
			"pinPadLowBattery": false,
			"lockDisabled": false,
			"lockLowBattery": false,
			"calibrationErrDelta": 0,
			"calibrationErrZero": 0,
			"lockJamState": 0
		},
		"name": "Front Door",
		"deviceGroupID": 1,
		"firmwareVersion": "1.0.0",
		"bootVersion": "1.0.0",
		"setting": {
			"autoLock": 3,
			"away": 1,
			"home": 1,
			"awayToOff": 0,
			"homeToOff": 1
		},
		"flags": {
			"swingerShutdown": false,
			"lowBattery": false,
			"offline": false
		}
	}, {
		"serial": "<device_serial>",
		"type": 253,
		"status": {},
		"name": "Front Door",
		"deviceGroupID": 1,
		"setting": {},
		"flags": {
			"swingerShutdown": false,
			"lowBattery": false,
			"offline": false
		}
	}],
	"lastUpdated": <unix_timestamp>,
	"lastSynced": <unix_timestamp>,
	"lastStatusUpdate": <unix_timestamp>
}```

I took out some personal information.

**Comment by bachya** [comment]
Thanks, @aceresia. A couple questions for you:

* You have two sensors, each named `Front Door`. Neither of their device type IDs (`16` and `253`) is ones that I've seen before. I'm assuming the first sensor is the lock; what is the second one?
* Have you captured the calls/traffic needed for actually locking and unlocking the lock?

**Comment by aceresia** [comment]
I believe 16 is the lock itself and 253 is the keypad that comes with the lock.

I'm still working on getting the calls to lock and unlock it.  I usually take a calls from the webapp but they're not present on that app, only on the mobile app.  I'll set up a MITM later this week to capture them.

**Comment by aceresia** [comment]
I got the commands for locking and unlocking.  When I captured it was HTTP2, but I believe I got the correct info.


POST https://api.simplisafe.com/v1/doorlock/<alarm_id>/<lock_serial>/state
content-type: application/json
Lock Payload: {"state":"lock"}
Unlock Payload: {"state":"unlock"}

**Comment by bachya** [comment]
What response payload comes after both calls?

**Comment by bachya** [comment]
@aceresia Thanks! I'm getting close to a version of this PR you can test. More tomorrow.

**Comment by aceresia** [comment]
The response to these requests is the same JSON as the payload, just with the new state.

after locking: {"state":"lock"}

**Comment by bachya** [comment]
Thanks @aceresia!

If you're willing, I think we have enough in this PR to begin testing. I've added a simple script that, at a minimum, should see if my assumptions around lock properties and state are correct.

## Testing Plan

Clone the repo and check out this PR:

```bash
$ git clone https://github.com/bachya/simplisafe-python.git
$ git checkout origin/lock
```

Initialize the dev environment:

```bash
$ make init
```

Open up `example_locks.py` in a text editor and input your SimpliSafe email/password into lines 13 and 14:

```python
SIMPLISAFE_EMAIL = "<EMAIL>"
SIMPLISAFE_PASSWORD = "<PASSWORD>"
```

Save the file and close your editor. Back in your terminal, run the script – note that based on current assumptions, this _should_:

1. Print out the serial number, name, and current state of each lock
2. If the lock is currently locked, unlock it, sleep for 3 seconds, and lock it again
2. Conversely, if the lock is currently unlocked, lock it, sleep for 3 seconds, and unlock it again

```bash
$ pipenv run python3 example_locks.py
```

## What I Need

Since I don't have a lock of my own to test, I am making a lot of assumptions, So, as you test, please provide two things:

1. Paste the output of the script's run here (removing any sensitive information as necessary)
2. Let me know if the lock actually does what it's supposed to (`Lock -> Unlock` or `Unlock -> Lock`, depending on the lock's initial state)

**Comment by aceresia** [comment]
So I tried it and it looks like I read wireshark wrong.  There actually isn't a return value.  So I commented out lines 63-66 of lock.py and it unlocked and then re-locked itself.

Output:
INFO:root:Lock <lock_serial>: (name: Front Door, state: LockStates.locked)
INFO:root:Unlocking...
doorlock/<alarm_id>/<lock_serial>/state
INFO:root:Locking...
doorlock/<alarm_id>/<lock_serial>/state

**Comment by bachya** [comment]
@aceresia Thanks! Glad to hear it worked. Any errors/incorrect behavior you could see?

**Comment by bachya** [comment]
Merged into `dev` (7722a33612c0). This will be a part of [5.1.0](https://github.com/bachya/simplisafe-python/releases/tag/untagged-8affff83c1dfe3882319), but I'm going to hold on releasing it while I diagnose [a possible issue](https://github.com/home-assistant/home-assistant/issues/23669#issuecomment-552123117) that could necessitate additional commits.

Thanks!
